flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/streaming/connectors/kafka/sink/KafkaProducerRecordSerializationSchema.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.connectors.kafka.sink;
+
+import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.api.common.serialization.SerializationSchema;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+
+import java.io.Serializable;
+
+/**
+ * A serialization schema which defines how to convert a value of type {@code T} to {@link
+ * ProducerRecord}.
+ *
+ * @param <T> the type of values being serialized
+ */
+@PublicEvolving
+public interface KafkaProducerRecordSerializationSchema<T> extends Serializable {
+
+    /**
+     * Initialization method for the schema. It is called before the actual working methods {@link
+     * #serialize(Object, FlinkKafkaSinkContext, Long)} and thus suitable for one time setup work.
+     *
+     * <p>The provided {@link SerializationSchema.InitializationContext} can be used to access
+     * additional features such as e.g. registering user metrics.
+     *
+     * @param context Contextual information that can be used during initialization.
+     */
+    default void open(SerializationSchema.InitializationContext context) throws Exception {}
+
+    /**
+     * Serializes given element and returns it as a {@link ProducerRecord}.
+     *
+     * @param element element to be serialized
+     * @param context context to possibly determine target partition
+     * @param timestamp timestamp
+     * @return Kafka {@link ProducerRecord}
+     */
+    ProducerRecord<byte[], byte[]> serialize(
+            T element, FlinkKafkaSinkContext context, Long timestamp);
+
+    /**
+     * Returns the target topic for the given record.
+     *
+     * @return topic id
+     */
+    String getTargetTopic(T record);
```

Comment:
Why do we need this? The user would set the topic directly on the `ProducerRecord`...